render-wasm/src/shapes/paths/subpaths.rs

```diff
@@ -17,230 +15,188 @@ impl Subpath {
+        })
     }
 
     pub fn is_closed(&self) -> bool {
         self.closed.unwrap_or_else(|| self.calculate_closed())
     }
 
-    pub fn add_segment(&mut self, segment: Segment) {
-        self.segments.push(segment);
-        self.closed = None;
-    }
-
     pub fn reversed(&self) -> Self {
-        let mut reversed = self.clone();
-        reversed.segments.reverse();
-        reversed
+        let mut rev = self.clone();
+        rev.segments.reverse();
+        rev.closed = None;
+        rev
     }
 
     fn calculate_closed(&self) -> bool {
         if self.segments.is_empty() {
-            return false;
+            return true;
         }
-
-        // Check if the path ends with a Close segment
         if let Some(Segment::Close) = self.segments.last() {
             return true;
         }
-
-        // Check if the first and last points are close to each other
-        if let (Some(first), Some(last)) = (self.segments.first(), self.segments.last()) {
-            let first_point = match first {
-                Segment::MoveTo(xy) => xy,
-                _ => return false,
-            };
-
-            let last_point = match last {
-                Segment::LineTo(xy) => xy,
-                Segment::CurveTo((_, _, xy)) => xy,
-                _ => return false,
-            };
-
-            return are_close_points(*first_point, *last_point);
+        if let (Some(first), Some(last)) = (self.start(), self.end()) {
+            return are_close_points(first, last);
         }
-
         false
     }
 }
 
-impl Default for Subpath {
-    fn default() -> Self {
-        Self::new(vec![])
-    }
+fn are_close_points(a: Point, b: Point) -> bool {
+    let tol = 1e-1;
+    (a.0 - b.0).abs() < tol && (a.1 - b.1).abs() < tol
+}
+
+#[derive(Debug, Clone)]
+enum MergeMode {
+    EndStart,
+    StartEnd,
+    EndEnd,
+    StartStart,
 }
 
-/// Joins two subpaths into a single subpath
 impl TryFrom<(&Subpath, &Subpath)> for Subpath {
-    type Error = String;
+    type Error = &'static str;
+    fn try_from((a, b): (&Subpath, &Subpath)) -> Result<Self, Self::Error> {
+        let mut segs = a.segments.clone();
+        segs.extend_from_slice(&b.segments);
+        Ok(Subpath::new(segs))
+    }
+}
 
-    fn try_from((subpath, other): (&Subpath, &Subpath)) -> Result<Self> {
-        if subpath.is_empty() || other.is_empty() || subpath.end() != other.start() {
-            return Err("Subpaths cannot be joined".to_string());
+pub fn closed_subpaths(subpaths: Vec<Subpath>) -> Vec<Subpath> {
+    let n = subpaths.len();
+    if n == 0 {
+        return vec![];
+    }
+
+    let mut used = vec![false; n];
+    let mut result = Vec::with_capacity(n);
+
+    for i in 0..n {
+        if used[i] {
+            continue;
         }
 
-        let mut segments = subpath.segments.clone();
-        segments.extend_from_slice(&other.segments);
-        Ok(Subpath::new(segments))
-    }
-}
+        let mut current = subpaths[i].clone();
```

Comment:
I'm trying to see why is necessary to clone this. I'm not seeing where the mutable reference.

Reply:
this is because `current` can be reassigned later